### ISSUE TITLE
Improved support for textDocument/publishDiagnostics

### ISSUE THREAD

**Issue by erlanger** [jamesnvc/lsp_server#4]
Contiunuing from #2 (LanguageClient)

The main functionality I was looking for was to show the compilation errors with textDocument/publishDiagnostics; is that supported?

If not, I have some code  from someone that tried to implement the lsp server (in the new discourse forum) and it collects the compilation errors.

**Comment by jamesnvc** [comment]
It isn't yet; I will try to do that next.

**Comment by erlanger** [comment]
Great! Look forward to it!

**Comment by jamesnvc** [comment]
Deployed v1.3.0 with preliminary support for publishDiagnostics.

**Comment by erlanger** [comment]
I am not seeing any errors, most likely because didChange is not being handled (LanguageClient sends the whole document again on didChange).

**Comment by jamesnvc** [comment]
Hm, I was able to see errors with LanguageClient; it does the error stuff in didSave, so the changes shouldn’t be a problem. I’ll have a look tomorrow; I also have some improvements to the warnings for singleton variables in progress.

**Comment by erlanger** [comment]
>  it does the error stuff in didSave, so the changes shouldn’t be a problem.

Ahh, what I meant was that if you did the error checking on didChange, it would work real time! But you probably have to do a threaded implementation to do that.

Thanks!
BTW, look at the update above about using `thread_message_hook/3` with `load_files/2` that way you can just gather all the errors produced by the compiler.

**Comment by jamesnvc** [comment]
Ah okay. I’d like to hook in to didChange at some point, but it isn’t working yet. 

Regarding “load_file”, right now the process is working via the xref library, which just uses static analysis. It does make getting a lot of the diagnostics more challenging, but I’m doing it this way because I’m not confident in how robust it will be if it tries to actually load the source files to check them (e.g. if initialization goals run, top-level queries might have side effects, code that redefine predicates could interfere with the checker itself).

**Comment by erlanger** [comment]
>  I’m not confident in how robust it will be if it tries to actually load the source files to check them

I see. I would ask Jan, I suspect there might be some way to get all the compiler errors in a safe way.

In any case, thanks for your great work!

**Comment by jamesnvc** [comment]
> I see. I would ask Jan, I suspect there might be some way to get all the compiler errors in a safe way.

Indeed. I’m about to go on vacation for two weeks; when I return, I’ll look into that. I’m also looking at the source for prolog_codewalk, to see if I can leverage that directly somehow. 

In the meantime, I’ve released 1.3.2, which should have better display of singleton variable warnings.

**Comment by erlanger** [comment]
Great! thanks!
